plugins/eos/gs-plugin-eos.c

```diff
@@ -1553,3 +1850,191 @@ gs_plugin_add_popular (GsPlugin *plugin,
+			     "background: url('" DATADIR "/gnome-software/upgrade-bg.png');"
+			     "background-size: 100% 100%;");
+
+	g_signal_connect_object (priv->updater_proxy, "notify::state",
+				 G_CALLBACK (updater_state_changed), plugin,
+				 G_CONNECT_SWAPPED);
+	g_signal_connect_object (priv->updater_proxy, "notify::downloaded-bytes",
+				 G_CALLBACK (updater_state_changed), plugin,
+				 G_CONNECT_SWAPPED);
+
+	priv->os_upgrade = app;
+
+	return app;
+}
+
+static gboolean
+should_add_os_upgrade (GsApp *os_upgrade)
+{
+	switch (gs_app_get_state (os_upgrade)) {
+	case AS_APP_STATE_AVAILABLE:
+	case AS_APP_STATE_INSTALLING:
+	case AS_APP_STATE_UPDATABLE:
+		return TRUE;
+	default:
+		break;
+	}
+
+	return FALSE;
+}
+
+gboolean
+gs_plugin_refresh (GsPlugin *plugin,
+		   guint cache_age,
+		   GsPluginRefreshFlags flags,
+		   GCancellable *cancellable,
+		   GError **error)
+{
+	g_autoptr(GError) local_error = NULL;
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+
+	/* if we already know we should add the OS upgrade, no need to look for
+	 * updates */
+	if (priv->os_upgrade && should_add_os_upgrade (priv->os_upgrade))
+		return TRUE;
+
+	/* poll so we eventually get a state change for the OS update, but never
+	 * return an error since we deal with them in the state change callback */
+	if (!eos_updater_call_poll_sync (priv->updater_proxy,
+					 cancellable, &local_error))
+		g_warning ("Error polling OS upgrade: %s", local_error->message);
+
+	return TRUE;
+}
+
+gboolean
+gs_plugin_add_distro_upgrades (GsPlugin *plugin,
+			       GsAppList *list,
+			       GCancellable *cancellable,
+			       GError **error)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	GsApp *os_upgrade = get_os_upgrade (plugin);
+
+	/* if we are testing the plugin, then always add the OS upgrade */
+	if (g_getenv ("GS_PLUGIN_EOS_TEST") != NULL) {
+		if  (gs_app_get_state (os_upgrade) == AS_APP_STATE_UNKNOWN)
+			gs_app_set_state (os_upgrade, AS_APP_STATE_AVAILABLE);
+		gs_app_list_add (list, os_upgrade);
+		return TRUE;
+	}
+
+	if (sync_state_from_updater (plugin) == EOS_UPDATER_STATE_READY) {
+		eos_updater_call_poll (priv->updater_proxy, cancellable, NULL,
+				       NULL);
+		return TRUE;
+	}
+
+	if (should_add_os_upgrade (os_upgrade)) {
+		g_debug ("Adding EOS upgrade: %s",
+			 gs_app_get_unique_id (os_upgrade));
+		gs_app_list_add (list, os_upgrade);
+	}
+
+	return TRUE;
+}
+
+gboolean
+gs_plugin_app_upgrade_download (GsPlugin *plugin,
+				GsApp *app,
+			        GCancellable *cancellable,
+				GError **error)
+{
+	g_autoptr(GError) local_error = NULL;
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	GsApp *os_upgrade = NULL;
+	gboolean needs_polling = FALSE;
+	EosUpdaterState eos_updater_state;
+
+	/* only process this app if was created by this plugin */
+	if (g_strcmp0 (gs_app_get_management_plugin (app),
+		       gs_plugin_get_name (plugin)) != 0)
+		return TRUE;
+
+	os_upgrade = get_os_upgrade (plugin);
+	if (app != os_upgrade) {
```

Comment:
Should we instead assert on this?

Reply:
I had an assert but changed it before pushing the PR because when the code is released we should ensure that we get some log if this happens + not let the logic proceed.